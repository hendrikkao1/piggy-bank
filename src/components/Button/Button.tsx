import React from "react";

interface ButtonProps<T extends React.ElementType> {
  as?: T;
  children?: React.ReactNode;
}

export const Button = <T extends React.ElementType = "button">({
  children,
  as,
  ...props
}: ButtonProps<T> &
  Omit<React.ComponentPropsWithoutRef<T>, keyof ButtonProps<T>>) => {
  const Component = as || "button";

  return (
    <Component
      className="text-sm font-semibold transition rounded-full bg-zinc-900 py-2 px-4 text-white hover:bg-zinc-600 "
      {...props}
    >
      {children}
    </Component>
  );
};
