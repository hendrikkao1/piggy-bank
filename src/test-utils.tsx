import React from "react";
import {
  RenderHookOptions,
  RenderOptions,
  render,
  renderHook,
} from "@testing-library/react";
import { NextIntlClientProvider } from "next-intl";
import messages from "./messages/en-US.json";

export const AllTheProviders = ({
  children,
}: {
  children: React.ReactNode;
}) => (
  <NextIntlClientProvider locale={"en-US"} messages={messages}>
    {children}
  </NextIntlClientProvider>
);

const customRender = (ui: React.ReactElement, options: RenderOptions = {}) =>
  render(ui, { wrapper: AllTheProviders, ...options });

const customRenderHook = <TProps, TResult>(
  callback: (props: TProps) => TResult,
  options?: RenderHookOptions<TProps>,
) => renderHook(callback, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";

export { customRender as render, customRenderHook as renderHook };
