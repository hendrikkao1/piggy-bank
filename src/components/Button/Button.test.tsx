import { render } from "@/test-utils";
import { Button } from "./Button";

describe("Button", () => {
  it("renders", () => {
    const { container } = render(<Button>Button</Button>);

    expect(container).toMatchSnapshot();
  });

  it("renders as link", () => {
    const { container } = render(
      <Button as="a" href="/link">
        Button
      </Button>,
    );

    expect(container).toMatchSnapshot();
  });

  it("accepts props", () => {
    const { container } = render(<Button disabled>Button</Button>);

    expect(container).toMatchSnapshot();
  });
});
