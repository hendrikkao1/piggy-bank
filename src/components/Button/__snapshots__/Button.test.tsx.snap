// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button accepts props 1`] = `
<div>
  <button
    class="text-sm font-semibold transition rounded-full bg-zinc-900 py-2 px-4 text-white hover:bg-zinc-600 "
    disabled=""
  >
    Button
  </button>
</div>
`;

exports[`Button renders 1`] = `
<div>
  <button
    class="text-sm font-semibold transition rounded-full bg-zinc-900 py-2 px-4 text-white hover:bg-zinc-600 "
  >
    Button
  </button>
</div>
`;

exports[`Button renders as link 1`] = `
<div>
  <a
    class="text-sm font-semibold transition rounded-full bg-zinc-900 py-2 px-4 text-white hover:bg-zinc-600 "
    href="/link"
  >
    Button
  </a>
</div>
`;
